languages/kotlin/generic/Parse_kotlin_tree_sitter.ml

```diff
@@ -2258,6 +2298,51 @@ let file_annotation (env : env) ((v1, v2, v3, v4, v5) : CST.file_annotation) =
   let _semi = semi env v5 in
   ()
 
+let merge_class_declarations (xs : stmt list) : stmt list =
+  let rec merge rev_acc (xs : stmt list) =
+    match xs with
+    (* Merge two consecutive class declarations, if the second one
+     * is a partial class declaration.
+     *
+     * For the first class declaration, only the name, kind of class,
+     * and attributes will be populated because it will come from the
+     * code
+     *   @SomeAttribute class SomeClassName
+     * or
+     *   @SomeAttribute enum SomeClassName
+     *
+     * The rest is populated by the second part which is a partial
+     * class declaration.
+     *)
+    | {
+        s = DefStmt ({ name; attrs; tparams = None }, ClassDef { ckind; _ });
+        _;
+      }
+      :: {
+           s =
+             DefStmt
+               ( {
+                   name = EN (Id (("@@@PARTIAL_CLASS_DECLARATION", _), _));
+                   attrs = [];
+                   tparams;
+                 },
+                 ClassDef
+                   { ckind = _; cextends; cimplements; cmixins; cparams; cbody }
+               );
+           _;
+         }
+      :: rest ->
+        let ent = { name; attrs; tparams } in
+        let cdef =
+          ClassDef { ckind; cextends; cimplements; cmixins; cparams; cbody }
+        in
+        let stmt = DefStmt (ent, cdef) |> G.s in
+        merge (stmt :: rev_acc) rest
```

Comment:
This implies that a given class declaration can be associated with only one following partial class declaration. Is that the case? If not, could you add a test case illustrating a class declaration with multiple following partial class declarations and adjust the implementation if needed?

Reply:
> given class declaration can be associated with only one following partial class declaration
Ah yes. Thanks for explicitly pointing out that assumption.

If we see a dangling partial class definition floating around, I'm pretty sure it's an invalid program. I tried to make a valid program that passes `kotlinc` but wasn't successful.  This part makes things look even a bit more hacky as I would expect parsing errors to be caught during the tree-sitter phase.

I should update the implementation to return a "parsing" error if there's a dangling partial class definition.

Reply:
Or just keep the dangling partial class definition and log a warning because it may be useful to match anyway. Hmm, I'm actually inclined to do this a bit more because it allows users to still find a match in the valid class declaration.

Reply:
I'd be cautious about erroring. In general we want to return our best effort at an AST even in the face of invalid programs (or programs that are grammar does not yet support).

I guess to rephrase my original question, is it valid Kotlin to have, for example, a class declaration of the sort you're testing here but with more members than just the constructor? If so, could you test that?

Reply:
Got it. No, other members will only appear in the class body (in braces), but still I think it's a good idea to make sure those cases are covered.

And I agree with you. Best effort AST sounds more reasonable than erroring here.

Reply:
Added more comments about this assumption and how we attempt to handle cases we don't expect.

Reply:
And added another test case that touches more complicated declarations.